presto-spi/src/main/java/com/facebook/presto/spi/relation/LogicalRowExpressions.java

```diff
@@ -384,14 +442,64 @@ private RowExpression notCallExpression(RowExpression argument)
         {
             return new CallExpression("not", functionResolution.notFunction(), BOOLEAN, singletonList(argument));
         }
+    }
 
-        private boolean isConjunctionOrDisjunction(RowExpression expression)
+    private class ConvertNormativeVisitor
+            implements RowExpressionVisitor<RowExpression, Form>
+    {
+        @Override
+        public RowExpression visitSpecialForm(SpecialFormExpression specialForm, Form clauseJoiner)
         {
-            if (expression instanceof SpecialFormExpression) {
-                Form form = ((SpecialFormExpression) expression).getForm();
-                return form == AND || form == OR;
+            if (!isConjunctionOrDisjunction(specialForm)) {
+                return specialForm;
             }
-            return false;
+
+            // normalize arguments
+            RowExpression left = specialForm.getArguments().get(0).accept(new ConvertNormativeVisitor(), clauseJoiner);
+            RowExpression right = specialForm.getArguments().get(1).accept(new ConvertNormativeVisitor(), clauseJoiner);
+
+            // check if already in correct form
+            if (specialForm.getForm() == clauseJoiner) {
+                return clauseJoiner == AND ? and(left, right) : or(left, right);
+            }
+
+            // else, we expand and distribute
```

Comment:
@wenleix, could you also help taking a look at this logic?

Reply:
@highker : I think an explanation about this algorithm is needed 😃